feat: mock_call with dynamic return data

ptisserand opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Closes #2861 

## Introduced changes

<!-- A brief description of the changes -->

- Add `mock_call_when`, `start_mock_call_when` and `stop_mock_call_when` to support "dynamic" return value  when mocking a contract entry point.

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

ptisserand committed:
feat(snforge_std): add "dynamic" return value for mock_call

ptisserand committed:
feat(cheatnet): add "dynamic" return value fir mock_call

ptisserand committed:
forge: update test to support new mock_call cheatcode

ptisserand committed:
fix mock_call test

ptisserand committed:
cheatnet: fixup clippy

franciszekjob commented on snforge_std/src/cheatcodes.cairo:
@cptartur I agree with this approach, we won't have breaking change + users won't be forced to pass `MockCallData::Any` frequently (which would happen in case of having current impl as `mock_call`).

ptisserand replied:
Clear, source code updated to remove the breaking change

ptisserand committed:
snforge_std: revert breaking changes on `mock_call`

Add `mock_call_when`, `start_mock_call_when` and `stop_mock_call_when`

ptisserand committed:
fix(cheatnet): default to any calldata entry when specific CheatSpan::TargetCalls is 0

ptisserand committed:
docs: add `mock_call_when` cheatcode

cptartur commented on crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/execution/entry_point.rs:
```diff
@@ -268,11 +270,25 @@ fn get_mocked_function_cheat_status<'a>(
+            let key_zero = (call.entry_point_selector, Felt::zero());
+
+            match contract_functions.get(&key) {
+                Some(CheatStatus::Cheated(_, CheatSpan::TargetCalls(0))) => {
```
After calling `get_mocked_function_cheat_status`, we call `.decrement_cheat_span` on the returned value. It changes the CheatSpan to `Uncheated` after it decreases to `0`. 

I don't think there can be a case where `TargetCalls` is equal to 0 unless explicitly set to this value somewhere.

cptartur replied:
I assume this logic match arm was added to handle the case where entrypoint for specific calldata is no longer cheated but for any calldata still is, but it should already be handled by `_` case anyway.

cptartur replied:
Turns out technically it could be set to `0` but it shouldn't be allowed.

Created #2927

ptisserand replied:
I'm not sure to understand what should I do here: remove my workaround?

cptartur replied:
You don't have to match for the case where cheat span is `CheatSpan::TargetCalls(0)`. Technically, user could have somehow created it like this (as explained in #2927), but during normal operation it will either be `CheatSpan::TargetCalls(some value > 0)` or `CheatStatus::Uncheated`.

cptartur replied:
See this https://github.com/foundry-rs/starknet-foundry/blob/f2e7d0d4077a654992914a688575a85bbe57b9a9/crates/cheatnet/src/state.rs#L141

ptisserand replied:
Workaround removed

cptartur commented on crates/cheatnet/src/runtime_extensions/forge_runtime_extension/cheatcodes/mock_call.rs:
```diff
@@ -9,26 +11,30 @@ impl CheatnetState {
-            function_selector,
-            CheatStatus::Cheated(ret_data.to_vec(), span),
-        );
+        let call_data_hash = match call_data {
```
Nitpick, but in most places in the code we already call it `calldata`, not `call_data`.

ptisserand replied:
Source code updated to use `calldata` instead of `call data`

cptartur commented on crates/forge/tests/integration/mock_call.rs:
I think we should also have test cases where `mock_call_when` is mixed with `mock_call`, these are the ones most likely to break/have some weird edge cases.

cptartur commented on docs/src/appendix/cheatcodes/mock_call_when.md:
```diff
@@ -0,0 +1,43 @@
+>   contract_address: ContractAddress, function_selector: felt252, call_data: MockCallData, ret_data: T, n_times: u32
+> )`
+
+Mocks contract call to a `function_selector` of a contract at the given address, with the given call data, for `n_times` first calls that are made 
```
Here and in other places

```suggestion
Mocks contract call to a `function_selector` of a contract at the given address, with the given calldata, for `n_times` first calls that are made 
```

ptisserand replied:
Source code updated to use `calldata` instead of `call data`

cptartur commented on snforge_std/src/cheatcodes.cairo:
```diff
@@ -20,6 +20,36 @@ pub enum CheatSpan {
     TargetCalls: usize,
 }
 
+/// Enum used to specify the call data that should be matched when mocking a contract call.
+#[derive(Copy, Drop, PartialEq, Clone, Debug)]
+pub enum MockCallData {
+    /// Matches any call data.
+    Any,
+    /// Matches the specified serialized call data.
+    Values: Span<felt252>,
+}
+
+impl MockCallDataSerde of Serde<MockCallData> {
+    fn deserialize(ref serialized: Span<felt252>) -> Option<MockCallData> {
+        let value: Option<Option<Span<felt252>>> = Serde::deserialize(ref serialized);
+
+        match value {
+            Option::None => Option::None,
+            Option::Some(call_data) => match call_data {
+                Option::None => Option::Some(MockCallData::Any),
+                Option::Some(data) => Option::Some(MockCallData::Values(data)),
+            },
+        }
+    }
+
+    fn serialize(self: @MockCallData, ref output: Array<felt252>) {
+        match self {
+            MockCallData::Any => Option::<Span<felt252>>::None.serialize(ref output),
+            MockCallData::Values(data) => Option::Some(*data).serialize(ref output),
+        }
+    }
+}
+
```
Why not just derive it?

```suggestion
/// Enum used to specify the call data that should be matched when mocking a contract call.
#[derive(Copy, Drop, PartialEq, Clone, Debug, Serde)]
pub enum MockCallData {
    /// Matches any call data.
    Any,
    /// Matches the specified serialized call data.
    Values: Span<felt252>,
}
```

ptisserand replied:
Source code updated in cheatnet crate introduce `MockCalldata` enum

ptisserand committed:
use `calldata` instead of `call_data`

ptisserand committed:
cheatnet: add MockCalldata enum

- rename cairo MockCallData enum into MockCalldata
- use Serde derivation for MockCalldata instead of custom serializer.

cptartur commented on crates/forge/tests/integration/mock_call.rs:
```diff
@@ -207,3 +207,376 @@ fn mock_calls() {
+}
+
+#[test]
+fn mock_calls_when_mixed() {
```
Can we add some cases where the same tests uses both `mock_call` and `mock_call_when`?

E.g. in case where you do `start_mock_call_when(MockCalldata::Any, ...)` and then `stop_mock_call(...)` and similar?

franciszekjob replied:
How about this @ptisserand ?

franciszekjob commented:
Hey @ptisserand 👋 , are you going to continue working on this PR?

ptisserand commented:
> Hey @ptisserand 👋 , are you going to continue working on this PR?

Hi @franciszekjob, sorry I completely forgot about this PR when working on other projects :(
If it's not blocking on your side, I can work on it in the incoming week: what do you think ?

franciszekjob commented:
@ptisserand sure thing, no hurry!

ptisserand committed:
Merge remote-tracking branch 'upstream/master'

ptisserand committed:
fix: add missing argument for run_test_case

ptisserand committed:
scarb fmt

franciszekjob commented:
Hi @ptisserand 👋 , kindly asking if there's any update on this?

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into feat/dynamic-mock-call

ptisserand committed:
remove workaround for #2927

cptartur commented:
Hey @ptisserand is this ready for re-review?

ptisserand commented:
> Hey @ptisserand is this ready for re-review?

Not yet, I have only managed conflicts with master branch.
I didn't yet add requested additional test cases.

franciszekjob commented:
Hi @ptisserand 👋 , any update on this?

ptisserand committed:
add test cases with interleaved start/stop mock_call and mock_call_when

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into feat/dynamic-mock-call

ptisserand committed:
cargo fmt

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into master-merged

ptisserand committed:
add missing use for cheatcodes

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into feat/dynamic-mock-call

franciszekjob commented on docs/src/appendix/cheatcodes/mock_call_when.md:
```diff
@@ -0,0 +1,43 @@
+See `mock_call_when` for comprehensive definition of how it can be used.
+
+
+### `stop_mock_call_when`
+
+> `fn stop_mock_call_when(contract_address: ContractAddress, function_selector: felt252, calldata: MockCalldata)`
```
```suggestion
## `stop_mock_call_when`
> `fn stop_mock_call_when(contract_address: ContractAddress, function_selector: felt252, calldata: MockCalldata)`
```

ptisserand replied:
Done

franciszekjob commented on CHANGELOG.md:
```diff
@@ -362,6 +362,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Rust is no longer required to use `snforge` if using Scarb >= 2.10.0 on supported platforms - precompiled `snforge_scarb_plugin` plugin binaries are now published to [package registry](https://scarbs.xyz) for new versions.
 - Added a suggestion for using the `--max-n-steps` flag when the Cairo VM returns the error: `Could not reach the end of the program. RunResources has no remaining steps`.
+- `mock_call_when`, `start_mock_call_when`, `stop_mock_call_when` cheatcodes.
```
It should be under unreleased version.

franciszekjob commented on docs/src/appendix/cheatcodes/mock_call_when.md:
```diff
@@ -0,0 +1,43 @@
+# `mock_call_when`
+
+Cheatcodes mocking contract entry point calls:
```
It's the same as description of `mock_call`. Maybe let's emphasize the difference?

ptisserand replied:
Updated with `Cheatcodes mocking contract entry point calls based on calldata:`

franciszekjob commented on docs/src/appendix/cheatcodes/mock_call_when.md:
It would be really nice to add an example of usage, similarly how it's done [here](https://foundry-rs.github.io/starknet-foundry/appendix/cheatcodes/mock_call.html#example) 😄 
You can create and issue for that, link it here and if you have time open - a separate PR.

ptisserand committed:
docs: apply PR suggestion

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into feat/dynamic-mock-call

ptisserand committed:
update CHANGELOG

ptisserand committed:
Merge remote-tracking branch 'upstream/master' into feat/dynamic-mock-call
